clients/src/main/java/org/apache/kafka/clients/consumer/internals/AssignmentReconciler.java

```diff
@@ -0,0 +1,263 @@
+        newAssignment.removeAll(revokedPartitions);
+        subscriptions.assignFromSubscribed(newAssignment);
+    }
+
+    /**
+     * Determine which partitions are newly revoked. This is done by comparing the
+     * {@link Assignment#topicPartitions() target set from the assignment} against the
+     * {@link SubscriptionState#assignedPartitions() current set}. The returned set of
+     * {@link TopicPartition partitions} are composed of any partitions that are in the current set but
+     * are no longer in the target set.
+     *
+     * @param assignment Holds the {@link Assignment} which includes the target set of topics
+     * @return Set of partitions to revoke
+     */
+
+    SortedSet<TopicPartition> getPartitionsToRevoke(Assignment assignment) {
+        SortedSet<TopicPartition> partitions = new TreeSet<>(new TopicPartitionComparator());
+        partitions.addAll(subscriptions.assignedPartitions());
+        partitions.removeAll(targetPartitions(assignment));
+        return partitions;
+    }
+
+    /**
+     * Determine which partitions are newly assigned. This is done by comparing the
+     * {@link Assignment#topicPartitions() target set from the assignment} against the
+     * {@link SubscriptionState#assignedPartitions() current set}. Any {@link TopicPartition partitions} from the
+     * target set that are not already in the current set are included in the returned set.
+     *
+     * @param assignment {@link Optional} that holds the {@link Assignment} which includes the target set of topics
+     * @return Set of partitions to assign
+     */
+
+    SortedSet<TopicPartition> getPartitionsToAssign(Assignment assignment) {
+        SortedSet<TopicPartition> partitions = new TreeSet<>(new TopicPartitionComparator());
+        partitions.addAll(targetPartitions(assignment));
+        partitions.removeAll(subscriptions.assignedPartitions());
+        return partitions;
+    }
+
+    /**
+     * Determine which partitions should be "lost". This is simply the
+     * {@link SubscriptionState#assignedPartitions() current set} of {@link TopicPartition partitions}.
+     *
+     * @return Set of partitions to "lose"
+     */
+    SortedSet<TopicPartition> getPartitionsToLose() {
+        SortedSet<TopicPartition> partitions = new TreeSet<>(new TopicPartitionComparator());
+        partitions.addAll(subscriptions.assignedPartitions());
+        return partitions;
+    }
+
+    private SortedSet<TopicPartition> targetPartitions(Assignment assignment) {
+        Map<Uuid, String> topicIdToNameMap = createTopicIdToNameMapping();
+        SortedSet<TopicPartition> partitions = new TreeSet<>(new TopicPartitionComparator());
+
+        for (TopicPartitions topicPartitions : assignment.topicPartitions()) {
+            Uuid topicId = topicPartitions.topicId();
+            String topicName = Objects.requireNonNull(topicIdToNameMap.get(topicId), () -> String.format("No topic name was found in the metadata for topic ID %s", topicId));
```

Comment:
The metadata cache may not have the relevant metadata for newly assigned topics so this won't work. I think that we must request a metadata update for the newly assigned topics.

Reply:
☹️ Let me think about how that affects things here...

Reply:
For my own understanding, when will a consumer receive assignments for topics which he hasn't subscribed to a priori? Is this the case where the user is using regex-based subscriptions?

cc @lianetm @philipnee

Reply:
The client currently _assumes_ that it can start reconciling immediately upon receipt of the new assignment. In this case, we will need to a) catch the case where we're missing metadata, b) request the metadata update and stash the assignment, and c) attempt to restart the processing of the reconciliation.